ReactWindows/ReactNative/Modules/Image/ImageLoaderModule.cs

```diff
@@ -15,8 +48,382 @@ public override string Name
+            lock (_waitingListImage)
+            {
+                var image = sender as BitmapImage;
+                RemoveEventHandlers(image);
+
+                ImageReference(uriString, _waitingListImage[uriString].Count, false);
+
+                lock (_images)
+                {
+                    _images[uriString] = image;
+                }               
+
+                foreach (var item in _waitingListImage[uriString])
+                {                
+                    manager.ImageLoaded(item, true, uriString);
+                }
+
+                _waitingListImage.Remove(uriString);
+            }
+        }
+
+        private void RemoveEventHandlers(BitmapImage image)
+        {
+            if (image == null)
+            {
+                return;
+            }
+
+            lock(_imageFailedHandlers)
+            {
+                var imageFailedHandler = default(ExceptionRoutedEventHandler);
+                if (_imageFailedHandlers.TryGetValue(image, out imageFailedHandler))
+                {
+                    _imageFailedHandlers.Remove(image);
+                    image.ImageFailed -= imageFailedHandler;
+                }
+            }
+            
+            lock(_imageOpenedHandlers)
+            {
+                var imageOpenedHandler = default(RoutedEventHandler);
+                if (_imageOpenedHandlers.TryGetValue(image, out imageOpenedHandler))
+                {
+                    _imageOpenedHandlers.Remove(image);
+                    image.ImageOpened -= imageOpenedHandler;
+                }
+            }         
+        }
+
+        private bool Waiting(Dictionary<string, List<Tuple<Border, Color?, Color?>>> waitList, string uriString, object data)
+        {
+            lock (waitList)
+            {
+                var list = default(List<Tuple<Border, Color?, Color?>>);
+
+                if (waitList.TryGetValue(uriString, out list))
+                {
+                    list.Add((Tuple<Border, Color?, Color?>)data);
+                    return true;
+                }
+                else
+                {
+                    waitList.Add(
+                        uriString,
+                        new List<Tuple<Border, Color?, Color?>>()
+                        {
+                            (Tuple<Border, Color?, Color?>)data
+                        });
+
+                    return false;
+                }
+            }
+        }
+
+        private void ImageReference(string uriString, int count, bool prefetch = false)
+        {
+            if (uriString == null)
+            {
+                return;
+            }
+
+            lock (_imageReferences)
+            {
+                var reference = default(Tuple<int, bool>);
+
+                if (_imageReferences.TryGetValue(uriString, out reference))
+                {
+                    if (reference.Item1 + count == 0)
+                    {
+                        lock (_waitingListImage)
```

Comment:
Is there really going to be a lot of contention across all these dictionaries?  Couldn't we instead just use a single lock?